doc/manual/source/store/building.md

```diff
@@ -1,101 +1,187 @@
 
 - **Normalize the file permissions**
 
-  Nix sets the last-modified timestamp on all files
-  in the build result to 1 (00:00:01 1/1/1970 UTC), sets the group to
-  the default group, and sets the mode of the file to 0444 or 0555
-  (i.e., read-only, with execute permission enabled if the file was
-  originally executable). Any possible `setuid` and `setgid`
-  bits are cleared.
-
-  > **Note**
-  >
-  > Setuid and setgid programs are not currently supported by Nix.
-  > This is because the Nix archives used in deployment have no concept of ownership information,
-  > and because it makes the build result dependent on the user performing the build.
+  The files must conform to the model described in the [Exposing in OS file systems](./file-system-object/os-file-system.md) section.
+  For example, timestamps and permissions must be forced to sentinel values.
 
 - **Calculate the references**
 
-  Nix scans each output path for
-  references to input paths by looking for the hash parts of the input
-  paths. Since these are potential runtime dependencies, Nix registers
-  them as dependencies of the output paths.
+  Nix scans each output path for references to input store objects by looking for the store path digests of each input.
+  (The name part is ignored when scanning; an input's hash part that is not followed by a `-` and the correct name part still scans as a reference.
+  Likewise, a digest not preceded by the [store directory path] also still scans as a reference.)
+  Since these are potential runtime dependencies, Nix will register them as references of the output store object they occur in.
 
-  Nix also scans for references to other outputs' paths in the same way, because outputs are allowed to refer to each other.
+  Nix also scans for references from one output to another in the same way, because outputs are allowed to refer to each other.
   If the outputs' references to each other form a cycle, this is an error, because the references of store objects much be acyclic.
 
+  In the case of derivations with fixed in advance output paths (i.e. [input-addressing] derivations, or [fixed content-addressing] derivations), the actual final store path to each output is used during the build.
+  For [floating content-addressing] derivations, however, the final store path is not known in advance by definition.
+  Scratch store paths must therefore be used instead.
+  Scanning will use those scratch paths, but then any output-to-be that contains such a scanned scratch path must be rewritten to instead use the final (content-addressed) path of the output in question.
+
+At this point, the file system data is in the proper form, and the valid acyclic reference data for each output is also calculated, so the outputs can be registered as proper store objects, and associated with the derivation in the [build trace] in the record for a successful build.
```

Comment:
```suggestion
At this point, the file system data is in the proper form. Nix then computes the references for each output, so the outputs can be registered as proper store objects. References must form an acyclic graph.
```
(dropping the "build trace" since I don't know what that is)